src/custom/hooks/useWrapCallback.ts

```diff
@@ -0,0 +1,106 @@
+            } catch (error) {
+              console.error('Could not deposit', error)
+            }
+          }
+        : undefined,
+    inputError: sufficientBalance ? undefined : 'Insufficient ETH balance'
+  }
+}
+
+/**
+ * Given the selected input and output currency, return a wrap callback
+ * @param inputCurrency the selected input currency
+ * @param outputCurrency the selected output currency
+ * @param typedValue the user input value
+ */
+export default function useWrapCallback(
+  inputCurrency: Currency | undefined,
+  outputCurrency: Currency | undefined,
+  typedValue: string | undefined
+): { wrapType: WrapType; execute?: undefined | (() => Promise<void>); inputError?: string } {
+  const { chainId, account } = useActiveWeb3React()
+  const wethContract = useWETHContract()
+  const balance = useCurrencyBalance(account ?? undefined, inputCurrency)
+  const etherBalance = useCurrencyBalance(account ?? undefined, ETHER)
+  // we can always parse the amount typed as the input currency, since wrapping is 1:1
+  const inputAmount = useMemo(() => tryParseAmount(typedValue, inputCurrency), [inputCurrency, typedValue])
+  const addTransaction = useTransactionAdder()
+
+  return useMemo(() => {
+    if (!wethContract || !chainId || !inputCurrency || !outputCurrency) return NOT_APPLICABLE
+
+    const sufficientBalance = !!(inputAmount && balance && !balance.lessThan(inputAmount))
+    const sufficientEtherBalance = !!(inputAmount && etherBalance && !etherBalance.lessThan(inputAmount))
+
+    // SELLING ETH for WETH
+    // = WRAP
+    if (inputCurrency === ETHER && currencyEquals(WETH[chainId], outputCurrency)) {
+      return wrapEther({ sufficientBalance, inputAmount, wethContract, addTransaction })
+      // SELLING WETH for ETH
+      // = UNWRAP
+    } else if (currencyEquals(WETH[chainId], inputCurrency) && outputCurrency === ETHER) {
+      return {
+        wrapType: WrapType.UNWRAP,
+        execute:
+          sufficientBalance && inputAmount
+            ? async () => {
+                try {
+                  const txReceipt = await wethContract.withdraw(`0x${inputAmount.raw.toString(16)}`)
+                  addTransaction(txReceipt, { summary: `Unwrap ${inputAmount.toSignificant(6)} WETH to ETH` })
+                } catch (error) {
+                  console.error('Could not withdraw', error)
+                }
+              }
+            : undefined,
+        inputError: sufficientBalance ? undefined : 'Insufficient WETH balance'
+      }
+      // SELLING WETH FOR WHATEVER (NOT ETHER)
+      // AND INSUFFICIENT WETH BALANCE BUT ETHER BALANCE === SUFFICIENT
+    } else if (
+      // Input = WETH
+      currencyEquals(WETH[chainId], inputCurrency) &&
+      // User doesn't have enough for trade
+      !sufficientBalance &&
+      // User has ETHER balance and
+      // ETHER balance is NOT less than input amount
+      sufficientEtherBalance
```

Comment:
What about when `!sufficientBalance` AND `!sufficientEtherBalance`, but `suffiectSumOfBalances (WETH+ETH)`
We should allow to wrap only what is missing, not just suddenly wrap all ETH  user might have, when he is only missing maybe 0.5 WETH